### ISSUE TITLE
Unable to resolve 'os' with ESM module dependency

### ISSUE THREAD

**Issue by quinnjr** [webpack/webpack#14675]
# Bug report

This bug report relates to [this] StackOverflow question and a bug report filed on the Pisma repo: https://github.com/prisma/prisma/issues/10058

**What is the current behavior?**

Bundling a module made using the ESM module system with a 'node' protocol import does not correctly resolve 'os' when the package requires it.

```sh
Error: Cannot find module 'os'
    at webpackEmptyContext (/home/joseph/Projects/cs-portfolio/CIS3950/agineek-website/dist/agineek-website/server/main.js:511630:10)
    at ../../node_modules/.pnpm/supports-color@7.2.0/node_modules/supports-color/index.js (/home/joseph/Projects/cs-portfolio/CIS3950/agineek-website/dist/agineek-website/server/main.js:193372:15)
    at __require2 (/home/joseph/Projects/cs-portfolio/CIS3950/agineek-website/dist/agineek-website/server/main.js:191265:45)
    at ../../node_modules/.pnpm/chalk@4.1.2/node_modules/chalk/source/index.js (/home/joseph/Projects/cs-portfolio/CIS3950/agineek-website/dist/agineek-website/server/main.js:193715:9)
    at __require2 (/home/joseph/Projects/cs-portfolio/CIS3950/agineek-website/dist/agineek-website/server/main.js:191265:45)
    at Object.59062 (/home/joseph/Projects/cs-portfolio/CIS3950/agineek-website/dist/agineek-website/server/main.js:231640:33)
    at __webpack_require__ (/home/joseph/Projects/cs-portfolio/CIS3950/agineek-website/dist/agineek-website/server/main.js:534735:42)
    at Object.76345 (/home/joseph/Projects/cs-portfolio/CIS3950/agineek-website/dist/agineek-website/server/main.js:190859:5)
    at __webpack_require__ (/home/joseph/Projects/cs-portfolio/CIS3950/agineek-website/dist/agineek-website/server/main.js:534735:42)
    at Object.28999 (/home/joseph/Projects/cs-portfolio/CIS3950/agineek-website/dist/agineek-website/server/main.js:191103:16) {
  code: 'MODULE_NOT_FOUND'
}
```

The present error results from, in my case, the module [`prisma`](https://github.com/prisma/prisma/) which depends on [`supports-color`](https://github.com/chalk/supports-color)

Relevant portions of the outputted bundle:

```js
var require_supports_color2 = __commonJS2({
  "../../node_modules/.pnpm/supports-color@7.2.0/node_modules/supports-color/index.js"(exports2, module2) {
    "use strict";

    var os2 = __require("os");

    var tty = __require("tty");

    var hasFlag = require_has_flag2();
    var {
      env
    } = process;
    var forceColor;
...
```

The issue is reported in the `prisma` repo as hitting both NestJS, NextJS, and NX workspace applications.

**If the current behavior is a bug, please provide the steps to reproduce.**

Minimal reproduction repos:

https://github.com/quinnjr/webpack-os-error-issue

Steps to observer:

1. Install dependencies and setup Prisma correctly
2. Generate Prisma client
3. run `npm run build:ssr`
4. Wait for webpack to finish bundling frontend and backend.
5. Run `npm run serve:ssr`

https://github.com/tonivj5/prisma-issue-10058

**What is the expected behavior?**

The expected behavior is for a Webpack bundle set with a node target to be able to resolve the `os` module.

**Other relevant information:**
webpack version:  v5.62.1
Node.js version: v16.13.0
Operating System: ArchLinux
Additional tools: `@angular/custom-webpack` @ ^12.1.3

**Comment by alexander-akait** [comment]
The same https://github.com/webpack/webpack/issues/14166, you should use `externals` right now to ignore or using own module, `node:` is protocol, so alising doesn't work, I want to close it because it is the same problem, but with different effects

**Comment by alexander-akait** [comment]
Use `externals: { "node:os": "{}" }`, but anyway it is bug and should work, closing in favor https://github.com/webpack/webpack/issues/14166

**Comment by quinnjr** [comment]
@alexander-akait The suggested solution does not resolve the issue in the `@nestjs/ng-universal` minimal reproduction repo.

I'm currently waiting on the people who reported the same issue for NextJS applications and NX workspaces to test the solution in their applications to see if it's an Angular-only issue.

See: https://github.com/prisma/prisma/issues/10058#issuecomment-963173043

**Comment by alexander-akait** [comment]
You need to create alias on real `os` module, `{}` just create empty modules, but I think you want to use `os` when you use `target: 'node'`, try:
```
'node.os': 'commonjs2 os',
```

**Comment by vankop** [comment]
I `target: node` you can use just `require('os')`, sounds like target is not a node..

**Comment by quinnjr** [comment]
@alexander-akait Still does not work in the Angular reproduction repo.

The author who reported issues in the NX project repo also reports back that the given solution does not work for his repo either.

@vankop `target = 'node'` is set in the configuration in the reproduction repos.

**Comment by alexander-akait** [comment]
It means no solution right now, you can wait a fix or help us and send a PR. 

Also in future please provide when you write:
> Install dependencies and setup Prisma correctly
> Generate Prisma client

unfortunately not every work with prisma and understand how to do it.

Anyway I can't reproduce it using pure setup, I think something wrong in your configuration or angular, unfortunately repo is very big. My pure configuration:
```
{
  mode: "production",
  devtool: false,
  entry: "./entry.mjs",
  target: 'node',
}
```

code:
```
import os from "node:os";

console.log(os);
```

And all works fine.

**Comment by quinnjr** [comment]
@alexander-akait I should also mention that I've reviewed the code after excluding `supports-color` from the bundle and the only issue seems to be including `node:os` in the final bundle. Other `node` protocol imports are working correctly (eg: tty, process)

**Comment by alexander-akait** [comment]
@quinnjr Please read https://github.com/prisma/prisma/issues/10058#issuecomment-964117671
